packages/@lwc/state/src/index.ts

```diff
@@ -1 +1,165 @@
+  private dependencies: Record<string, Signal<unknown>>;
+  private _value: T;
+
+  constructor(inputSignalsObj: Record<string, Signal<unknown>>, computer: Computer<unknown>) {
+    super();
+    this.computer = computer;
+    this.dependencies = inputSignalsObj;
+
+    const onUpdate = () => this.notify();
+    for (const signal of Object.values(inputSignalsObj)) {
+      signal.subscribe(onUpdate);
+    }
+  }
+
+  private computeValue() {
+    const dependencyValues: Record<string, unknown> = {};
+    for (const [signalName, signal] of Object.entries(this.dependencies)) {
+      dependencyValues[signalName] = signal.value;
+    }
+    this._value = this.computer(dependencyValues) as T;
+  }
+
+  protected override notify(): void {
+    this.computeValue();
+    super.notify();
+  }
+
+  get value() {
+    if (!this._value) {
+      this.computeValue();
+    }
+    return this._value;
+  }
+}
+
+const isUpdater = (signalOrUpdater: Signal<unknown> | ExposedUpdater) =>
+  typeof signalOrUpdater === 'function';
+
+const atom: MakeAtom = <T,>(initialValue: T) => new AtomSignal<T>(initialValue);
+
+const computed: MakeComputed = (inputSignalsObj, computer) =>
+  new ComputedSignal(inputSignalsObj, computer);
+
+const update: MakeUpdate<AtomSignal<unknown>> = <
+  AdditionalArguments extends unknown[],
+  Updater extends (signalValues: ValuesObj, ...args: AdditionalArguments) => ValuesObj,
+  SignalsObj extends Record<string, AtomSignal<unknown>>,
+  ValuesObj extends {
+    [signalName in keyof SignalsObj]?: UnwrapSignal<SignalsObj[keyof SignalsObj]>;
+  },
+>(
+  signalsToUpdate: SignalsObj,
+  updater: Updater,
+) => {
+  return (...uniqueArgs: AdditionalArguments) => {
+    const signalValues = {} as ValuesObj;
+
+    for (const [signalName, signal] of Object.entries(signalsToUpdate)) {
+      signalValues[signalName as keyof ValuesObj] = signal.value as ValuesObj[keyof ValuesObj];
+    }
+
+    const newValues = updater(signalValues, ...uniqueArgs);
+
+    for (const [atomName, newValue] of Object.entries(newValues)) {
+      signalsToUpdate[atomName][atomSetter](newValue);
+    }
+  };
+};
+
+export const defineState: DefineState = (defineStateCallback) => {
+  return (...args) => {
+    class StateManagerSignal<OuterStateShape> extends SignalBaseClass<OuterStateShape> {
+      private internalStateShape: Record<string, Signal<unknown> | ExposedUpdater>;
+      private isNotifyScheduled = false;
+      private _value: OuterStateShape;
+
+      constructor() {
+        super();
+
+        // @ts-ignore: TODO
+        const fromContext: MakeContextHook = () => {};
+
+        this.internalStateShape = defineStateCallback(atom, computed, update, fromContext)(...args);
+
+        for (const signalOrUpdater of Object.values(this.internalStateShape)) {
+          if (!isUpdater(signalOrUpdater)) {
+            // Subscribe to changes to exposed state atoms, so that the entire state manager signal
+            // "reacts" when the atoms change.
```

Comment:
```suggestion
            // Subscribe to changes to exposed state atoms and computeds, so that the entire
            //  state manager signal "reacts" when the atoms change.
```